src/server/scripts/EasternKingdoms/ZulGurub/boss_thekal.cpp

```diff
@@ -270,147 +353,135 @@ class boss_thekal : public CreatureScript
+                    Talk(TALK_FAKE_DEATH);
+                    me->RemoveAllAuras();
+                    me->SetImmuneToPC(true, true);
+                    me->SetImmuneToNPC(true, true);
+                    DoCastSelf(SPELL_PERMANENT_FEIGN_DEATH, true);
+                    me->AttackStop();
+                    if (Creature* thekal = _instance->GetCreature(DATA_THEKAL))
+                        thekal->AI()->SetData(DATA_FAKE_DEATH, me->GetEntry());
+                    damage = 0;
+                }
             }
 
-            void Reset() override
+            void JustEngagedWith(Unit* /*who*/) override
             {
-                Initialize();
-
-                _instance->SetBossState(DATA_LORKHAN, NOT_STARTED);
-                me->SetUInt32Value(UNIT_FIELD_BYTES_1, 0);
-                me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
+                _events.ScheduleEvent(EVENT_SHIELD, 1s);
+                _events.ScheduleEvent(EVENT_BLOODLUST, 16s);
+                _events.ScheduleEvent(EVENT_GREATER_HEAL, 32s);
+                _events.ScheduleEvent(EVENT_DISARM, 6s);
             }
 
             void UpdateAI(uint32 diff) override
             {
                 if (!UpdateVictim())
                     return;
 
-                if (_shieldTimer <= diff)
-                {
-                    DoCast(me, SPELL_SHIELD);
-                    _shieldTimer = 61000;
-                }
-                else
-                    _shieldTimer -= diff;
-
-                if (_bloodLustTimer <= diff)
-                {
-                    DoCast(me, SPELL_BLOODLUST);
-                    _bloodLustTimer = 20000 + rand32() % 8000;
-                }
-                else
-                    _bloodLustTimer -= diff;
-
-                // Casting Greaterheal to Thekal or Zath if they are in meele range.
-                if (_greaterHealTimer <= diff)
-                {
-                    Creature* thekal = _instance->GetCreature(DATA_THEKAL);
-                    Creature* zath = _instance->GetCreature(DATA_ZATH);
-                    if (!thekal && !zath)
-                        return;
-
-                    bool roll = roll_chance_i(50);
-                    Creature* target = roll ? (thekal ? thekal : zath) : (zath ? zath : thekal);
-                    // As we check already above to ensure at least 1 out of Thekal/Zath is not null, target must be not null
-                    ASSERT(target);
-
-                    if (!me->IsWithinMeleeRange(target))
-                        target = roll ? zath : thekal;
-
-                    if (me->IsWithinMeleeRange(target))
-                        DoCast(target, SPELL_GREATERHEAL);
-
-                    _greaterHealTimer = 15000 + rand32() % 5000;
-                }
-                else
-                    _greaterHealTimer -= diff;
+                // Don't update AI if faking death
+                if (me->HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC))
```

Comment:
if is not updating AI, should not be above UpdateVictim() call?

Reply:
Mistake in the comment, should be "don't update target". It avoids the creatures evading when they are immune to pc/npc and feigning death, but allows them to evade when the players wipe.